File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    
    int totalTime = 0; // 경과시간
    vector<int> truckIn(truck_weights.size(), 0); // 트럭이 처음 다리에 진입한 시간
    int totalWeight = 0; // 현재 다리에 트럭 무게
    queue<int> q; // 다리에 있는 트럭 무게
    
    int idx = 0;
    while(idx < truck_weights.size()) {
        // 경과 시간 증가
        totalTime++;
        
        // 다리에 차가 빠져나갈 수 있으면
        if(!q.empty() && totalTime - truckIn[q.front()] >= bridge_length) {
            // 무게 감소
            totalWeight -= truck_weights[q.front()];
            q.pop();
        }
        
        // 다리에 차가 진입할 수 있으면
        if(totalWeight + truck_weights[idx] <= weight) {
            // 무게 증가
            totalWeight += truck_weights[idx];
            // 큐에 삽입
            q.push(idx);
            // 시간 업데이트
            truckIn[idx] = totalTime;
            // 다음 트럭으로 이동
            idx++;
        }
    }
    
    // 마지막 트럭 시간 고려
    return totalTime + bridge_length;
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        
        // 큐 무게, [] 무게, 시작 시점, 현재 시간
        
        int time = 0;
        int bridge_weight=0;
        Queue<int[]> q = new ArrayDeque<>();
        int finished = 0;
        int idx = 0;
        
        while(true){
            // 다 지났으면
            if(finished == truck_weights.length){
                break;
            }
            
            // 시간 증가
            time++;
            
            // 큐에서 빠질 것들 체크
            while(!q.isEmpty() && time - q.peek()[1] >= bridge_length){
                bridge_weight -= q.poll()[0];
                finished++;
            }
            
            // 트럭 진입
            if(idx < truck_weights.length && bridge_weight + truck_weights[idx] <= weight){
                bridge_weight += truck_weights[idx];
                q.offer(new int[]{truck_weights[idx++], time});
            }
            
            
        }
 
        return time;
    }
}